src/wayland/data_device/device.rs

```diff
@@ -0,0 +1,164 @@
+use std::{cell::RefCell, os::unix::prelude::RawFd};
+
+use slog::debug;
+use wayland_server::{
+    protocol::{
+        wl_data_device::{self, WlDataDevice},
+        wl_data_offer::WlDataOffer,
+        wl_data_source::WlDataSource,
+        wl_seat::WlSeat,
+        wl_surface::WlSurface,
+    },
+    DelegateDispatch, DelegateDispatchBase, Dispatch, Resource,
+};
+
+use crate::wayland::{
+    compositor,
+    data_device::seat_data::{SeatData, Selection},
+    seat::Seat,
+    Serial,
+};
+
+use super::{dnd_grab, DataDeviceHandler, DataDeviceState, DataOfferUserData};
+
+static DND_ICON_ROLE: &str = "dnd_icon";
+
+/// Events that are generated by interactions of the clients with the data device
+#[derive(Debug)]
+pub enum DataDeviceEvent<D> {
```

Comment:
Since we use a handler trait, these enum variants could just be turned into trait functions.